streaming/src/main/scala/org/apache/spark/streaming/rdd/TrackStateRDD.scala

```diff
@@ -0,0 +1,190 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.streaming.rdd
+
+import java.io.{IOException, ObjectInputStream, ObjectOutputStream}
+
+import scala.collection.mutable.ArrayBuffer
+import scala.reflect.ClassTag
+
+import org.apache.spark.rdd.{MapPartitionsRDD, RDD}
+import org.apache.spark.streaming.{StateImpl, State}
+import org.apache.spark.streaming.util.{EmptyStateMap, StateMap}
+import org.apache.spark.util.Utils
+import org.apache.spark._
+
+private[streaming] case class TrackStateRDDRecord[K, S, T](
+    var stateMap: StateMap[K, S], var emittedRecords: Seq[T]) {
+  /*
+  private def writeObject(outputStream: ObjectOutputStream): Unit = {
+    outputStream.writeObject(stateMap)
+    outputStream.writeInt(emittedRecords.size)
+    val iterator = emittedRecords.iterator
+    while(iterator.hasNext) {
+      outputStream.writeObject(iterator.next)
+    }
+  }
+
+  private def readObject(inputStream: ObjectInputStream): Unit = {
+    stateMap = inputStream.readObject().asInstanceOf[StateMap[K, S]]
+    val numEmittedRecords = inputStream.readInt()
+    val array = new Array[T](numEmittedRecords)
+    var i = 0
+    while(i < numEmittedRecords) {
+      array(i) = inputStream.readObject().asInstanceOf[T]
+    }
+    emittedRecords = array.toSeq
+  }*/
```

Comment:
Still a bit of WIP. I wanted to optimized the serialization a little bit, but I will move this work to a different PR.